piccolo_admin/endpoints.py

```diff
@@ -214,6 +256,54 @@ def get_meta(self, request: Request) -> MetaResponseModel:
 
     ###########################################################################
 
+    def get_forms(self, request: Request) -> t.List[str]:
+        """
+        Returns a list of all forms from FormConfig.
+        """
+        return [form.name for form in self.forms]
+
+    def get_single_form_schema(self, request: Request, name: str) -> BaseModel:
+        param = request.path_params["name"]
+        for form in self.forms:
+            if form.name == param:
+                return form.pydantic_model.schema()
+
+    # we can add this method to dynamically change endpoints for axios.post
+    # def get_single_form_endpoint(self, request: Request, name: str) -> str:
+    #     param = request.path_params["name"]
+    #     for form in self.forms:
+    #         print(getattr(form.endpoint, "__name__", repr(form.endpoint)))
+    #         if form.name == param:
+    #             result = getattr(
+    #                 form.endpoint, "__name__", repr(form.endpoint)
+    #             )
+    #     return result
+
+    # manually processing the email form but no another form
+    async def post_single_form(self, request: Request) -> JSONResponse:
+        form = await request.json()
+
+        sender = "from@fromdomain.com"
+        receivers = [form["email"]]
+
+        message = f"""From: From Person <from@fromdomain.com>
+        To: To Person <{form["email"]}>
+        Subject: {form["title"]}
+
+        {form["content"]}
+        """
+
+        try:
+            smtpObj = smtplib.SMTP("localhost:1025")
+            smtpObj.sendmail(sender, receivers, message)
+            print("Successfully sent email")
+        except smtplib.SMTPException:
+            print("Error: unable to send email")
+
+        return JSONResponse({"form": form})
```

Comment:
Yeah. But this is the biggest problem as I already wrote yesterday. I don't know how call the different handler (end user wrote function) on the `FormConfig`. If you have any idea that would be great.

Reply:
@sinisaos I think modify ``FormConfig``, so it has a slug property:

```python
@dataclass
class FormConfig:
    name: str
    pydantic_model: t.Type[BaseModel]
    endpoint: t.Callable

    def __post_init__(self):
        self.slug = self.name.replace(' ', '-').lower()

```

Then in `__init__` for `AdminRouter` do something like this:

```python
self.form_config_map = {form.slug: form for form in forms}
```

Within `post_single_form` you can then do something like this:

```python
form_name = request.path_params["name"]
form_config = self.form_config_map.get(form_name)

data = await request.json()

try:
    model_instance = form_config.pydantic_model(**data)
except ValidationError as exception:
    return JSONResponse({'message': exception.message}, status_code=400)

try:
    form_config.endpoint(model_instance)
except ValidationError as exception:
    return JSONResponse({'message': exception.message}, status_code=400)

return JSONResponse({'message', 'Successfully submitted'})
```

Something like this should work in theory ...

Reply:
Well done! This work like a charm :). I slightly modified your code to also pass form data to endpoint.
```python
async def post_single_form(self, request: Request) -> JSONResponse:
    form_name = request.path_params["name"]
    form_config = self.form_config_map.get(  
        form_name.replace(" ", "-").lower()  <-- to get correct form name
    )
    data = await request.json()

    try:
        model_instance = form_config.pydantic_model(**data)
    except ValidationError as exception:
        return JSONResponse(
            {"message": exception.message}, status_code=400
        )

    try:
        form_config.endpoint(model_instance, data) <-- pass data to handler
    except ValidationError as exception:
        return JSONResponse(
            {"message": exception.message}, status_code=400
        )

    return JSONResponse({"message": "Successfully submitted"})
```
And result is this. Mail is sent from two different forms.

![forms](https://user-images.githubusercontent.com/30960668/132960663-cef2b345-c434-4b81-b841-c91243d59fca.png)

Thanks for this. Tomorrow I'll make commit with this changes.

Reply:
@sinisaos  Great, thanks! It's looking good.

I had one idea - we don't have to add it now, but it would be cool if the `FormConfig` object had a description attribute too which we would display in the UI. Not important for now, but maybe in the future.